### ISSUE TITLE
Roadmap for supporting wide-gamut color workflows

### ISSUE THREAD

**Issue by donmccurdy** [mrdoob/three.js#26479]
### Description

Related:

- https://github.com/mrdoob/three.js/issues/23614

Roadmap and discussion for supporting for wide-gamut color spaces in three.js. Many displays now support Display P3 (+50% larger gamut volume than sRGB), and web browsers are beginning to extend that support to WebGL contexts. Other color wide-gamut color spaces exist (Rec. 2020, ACEScg) and may be relevant to our internal rendering pipeline, but probably will not be widely available as *display* color spaces in WebGL for a very long time.

Broadly the goals here are (1) to identify useful and correct workflows for users, and (2) to make any necessary changes in the three.js rendering pipeline. I suspect that (1) may be the harder part of the problem.

## Decisions

- [ ] **Working color space:** In addition to "Linear sRGB", which working color space(s) should be supported?
    - Options include Linear P3, Linear Rec. 2020, and ACEScg. We may want to consider providing a mechanism to register new color spaces with three.js, rather than including conversions to/from each combination of color spaces out of the box. Each will also have different luminance calculations, different tone mapping, ...
- [ ] **Tone mapping:** What tone mapping should be used for wide-gamut displays?
- [ ] **Unlit vs Lit:** Support for unlit workflows (e.g. a photo editing app) will be easier than lit workflows, which require changes to tone mapping and perhaps PBR equations. Should we consider shipping unlit support first?

## Workflows

I'm skeptical than wide-gamut features can be enabled automatically; it will require significant care from users to get the benefits of these displays, and should be opt-in. I'm not sure which 3D modeling tools support the P3 gamut at all. Files in standard formats like glTF are almost exclusively using sRGB today. Users will need to load textures manually and assign `texture.colorSpace` correctly. We cannot infer texture color space safely without user input of some kind.

## Tasks

_Task numbers begin at 2.1, following up from #23614._

1. [ ] **(2.1)** Support changing `ColorManagement.workingColorSpace`
    - Does this affect `gl.unpackColorSpace`?
2. [ ] **(2.2)** Support THREE.Color getters/setters in Display P3 color space
4. [ ] **(2.2)** Support `renderer.outputColorSpace = THREE.DisplayP3ColorSpace`

### Solution

n/a

### Alternatives

n/a

**Comment by donmccurdy** [comment]
@WestLangley – continuing from our technical discussion, I've added more TODOs in the task list above. Please let me know if you feel anything major is missing or doesn't belong. :)

**Comment by donmccurdy** [comment]
PR for 2.1, 2.2, 2.3, and 2.6 —

- https://github.com/mrdoob/three.js/pull/26644

***

We should work toward offering a tone map adapted for wide-gamut and HDR displays, and with a more neutral default appearance than ACES Filmic. Blender 4.0 will be moving to AgX as the default, and I believe that would be a good choice for us as well. Related details:

- Implementation details: https://projects.blender.org/blender/blender/pulls/111099
- Screenshots: https://projects.blender.org/blender/blender/pulls/106355

Blender 4.0 alpha builds are available with AgX support now. I've prototyped an implementation in three.js. I am not sure about how to add support for other working and display color spaces yet. The implementation can also be optimized by removing the 1D LUT.

- https://glitch.com/edit/#!/three-sb2383?path=README.md%3A1%3A0

/cc @bhouston

**Comment by bhouston** [comment]
@donmccurdy great work!  I also strongly agree that we should aim to provide AgX as an alternative tone mapping and maybe if there is enough acceptance, we can switch it to the default.

**Comment by donmccurdy** [comment]
Updates:

- When drawing to a render target, we currently ignore the color space of the render target (WebGL will apply the sRGB OETF for us if necessary) and override outputColorSpace with LinearSRGBColorSpace. I believe this should semantically be the working color space, instead, but I'm not sure it affects the shader either way right now.
- Creating consistent color checker images in a variety of color spaces (sRGB, Linear sRGB, Display P3, Linear P3, ACEScg) has been challenging. OpenColorIO is not embedding the ICC Profiles we need for the web browser in PNG or JPEG, and so I'm waiting for a new alpha release of KTX-Software to get .ktx2 versions.
- In early tests with lit rendering in Linear Display P3, the most striking difference appears related to AO? If I light two scenes with a punctual light, they're very similar. If I light two scenes with an ambient light, they're similar only after disabling AO maps. The sRGB scene has much stronger AO. I've checked that the AO map (non-color data) is color managed correctly. Unsure why this is.
- It might be useful to have a utility function on KTX2Loader and EXRLoader (and anything else that supports RGBA8/16/32 uncompressed data) to assign or convert color spaces. Like `.assignColorSpace( SRGBColorSpace )` and `.convertColorSpace( LinearDisplayP3ColorSpace )`. Assign would override our beliefs about the existing color space of the data. Convert would convert from the existing color space to the given one. I wouldn't necessarily advise users to convert color spaces at runtime in production, but it will make development much easier.

**Comment by donmccurdy** [comment]
@WestLangley from your comment in https://github.com/mrdoob/three.js/pull/26666#discussion_r1356944154 –

> I don't think we should "honor" the color space of a render target when rendering to it. The color space is relevant only when reading from the render target.

If we're reading from the render target in sRGB we do need to get that data in correctly, right? We can safely "ignore" the color space now because writing Linear-sRGB to the sRGB target will lead WebGL to (automatically?) apply the sRGB OETF for us, and we don't want our fragment shader doing that.

I think we may want to go one step further here. If we are writing to a uint8 render target having Display P3 color space, from working color space Linear sRGB, our fragment shader should do the gamut mapping and let WebGL do the OETF. Writing non-linear data to a f16 or f32 render target would not be supported, because WebGL doesn't support it.

**Comment by WestLangley** [comment]
I think what you said is correct, and I don't think it is contrary to my statement. 🤔

**Comment by kitaedesigns** [comment]
Did this AgX feature request get merged into the latest version of threejs? I couldn't tell.  

> PR for 2.1, 2.2, 2.3, and 2.6 —
> 
> * [WebGLRenderer: Support "Linear Display P3" working color space and "Display P3" unlit rendering #26644](https://github.com/mrdoob/three.js/pull/26644)
> 
> We should work toward offering a tone map adapted for wide-gamut and HDR displays, and with a more neutral default appearance than ACES Filmic. Blender 4.0 will be moving to AgX as the default, and I believe that would be a good choice for us as well. Related details:
> 
> * Implementation details: https://projects.blender.org/blender/blender/pulls/111099
> * Screenshots: https://projects.blender.org/blender/blender/pulls/106355
> 
> Blender 4.0 alpha builds are available with AgX support now. I've prototyped an implementation in three.js. I am not sure about how to add support for other working and display color spaces yet. The implementation can also be optimized by removing the 1D LUT.
> 
> * https://glitch.com/edit/#!/three-sb2383?path=README.md%3A1%3A0
> 
> /cc @bhouston

**Comment by donmccurdy** [comment]
It hasn't been — I don't think anyone is working on it currently. Additional discussion here:

- https://github.com/mrdoob/three.js/issues/26661#issuecomment-1781228407

**Comment by kitaedesigns** [comment]
Gotcha thanks Don, I'm not sure your Glitch link is working anymore. With Blender 4.0 any assets users create in Blender and show in three.js look quite different. Your LUT approach although perhaps not totally accurate seems quite flexible and I think many 3D designers like myself are very used to LUTs due to their popularity in game engines like Unreal and Unity, would love to test with that LUT file, I assume it's a .CUBE or .png?

 I am currently using pmndrs PostProcessing library and @vanruesc mentioned it would be more efficient to have natively in three.js first which I understand, but maybe I can just use your LUT in his library and see how close I can match the colors between Blender 4 and Three.js.
